Adding new Symfony events for zipfile download and public link share

sharidas opened the pull request:
Adding new Symfony events, to check, if creating
shares or downloading zips are blocked by any
apps or not.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->

## Description
<!--- Describe your changes in detail -->
Adding new Symfony events to check if any apps revoked the premission to download zip files or creating public links.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
https://github.com/owncloud/core/issues/29877
https://github.com/owncloud/core/issues/29876

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Adding new symfony events which can be used by the apps to see if creating public links for files/folders are allowed or not. And also see if downloading zip files are allowed or not.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

sharidas committed:
Adding new Symfony events for zipfile download and public link share

Adding new Symfony events, to check, if creating
shares or downloading zips are blocked by any
apps or not.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented on lib/private/legacy/files.php:
```diff
@@ -139,6 +139,13 @@ public static function get($dir, $files, $params = null) {
 			}
 
+			//Dispatch an event to see if any apps have problem with download
+			$event = new \Symfony\Component\EventDispatcher\GenericEvent(null, ['files' => $files]);
```
Please double check what happens if "$dir" is set with a single value and multiple "$files" are set. It is likely that we also need to pass `$dir`. The alternative is to have a loop that prepends the directory to all given `$files` entries.

sharidas replied:
Here is the snapshot I took from IDE: 
![getmethod_singledir_multiplefiles](https://user-images.githubusercontent.com/3600427/34297966-3349a864-e741-11e7-8ec9-50896bb181cd.png)

Basically during testing I have used a directory and a file. Above is the screenshot of the `get` method with single dir `newfolder` and multiple files: `welcome.txt` and `rename_testfile.txt`. The newfolder and the files reside in the home folder of the user. The `$dir` is "/" in the above screenshot. I tried to create a folder inside `newfolder` and then created few files inside `newfolder`. In this case also `$dir` remains "/".

PVince81 commented on lib/private/legacy/files.php:
```diff
@@ -139,6 +139,13 @@ public static function get($dir, $files, $params = null) {
+			$event = new \Symfony\Component\EventDispatcher\GenericEvent(null, ['files' => $files]);
+			OC::$server->getEventDispatcher()->dispatch('file.beforecreatezip', $event);
+			if ($event->hasArgument('run') and ($event->getArgument('run') === false) and ($event->hasArgument('message'))) {
+				throw new Exception();
```
A generic exception without a message will confuse the user. Maybe throw ForbiddenException instead with a message "Access denied".

In general I'd rather expect event listeners to directly throw an exception from inside with a proper message, so the exception itself would already be forwarded by not catching it in the `dispatch()` call.

PVince81 commented on apps/files_sharing/lib/API/Share20OCS.php:
```diff
@@ -287,6 +293,15 @@ public function createShare() {
+		$event->setArgument('name', $name);
+		$event->setArgument('run', true);
+		//Dispatch an event to see if creating shares is blocked by any app
+		$this->eventDispatcher->dispatch('file.beforecreateShare', $event);
```
And also catch exception here and transform it to an OCS Result with the exception message. I expect that listeners will throw exceptions in case of denied access.

PVince81 commented on apps/files_sharing/lib/API/Share20OCS.php:
```diff
@@ -287,6 +293,15 @@ public function createShare() {
+		//Dispatch an event to see if creating shares is blocked by any app
+		$this->eventDispatcher->dispatch('file.beforecreateShare', $event);
+		if ($event->getArgument('run') === false) {
+			return new \OC\OCS\Result(null, 404, $this->l->t('No permission to create share'));
```
no permission is a 403 forbidden, not 404 not found

PVince81 reviewed (changes requested):
See comments above.

Note that for ZIP download I think there are at least 3 cases:

- URL with "files" with only one file => direct download
- URL with "files" and multiple files with full path => download all of them as ZIP
- URL with "dir" containing base dir and "files" containing a list of files inside that dir => prepend dir to all files and download as ZIP
- URL with "dir" without "files" => download said folder as ZIP

this is from memory, please check the code

sharidas commented:
Made the following changes:
- injected the event dispatcher to the constructor.
- used camel case for event name
- Replaced generic exception with ForbiddedException
- Added `$dir` to the event array
- Added after event to make sure the creation of share is successful.
@PVince81 Hope the new change looks better.

PVince81 commented on apps/files_sharing/lib/API/Share20OCS.php:
```diff
@@ -466,12 +484,15 @@ public function createShare() {
 
 		$share->getNode()->unlock(\OCP\Lock\ILockingProvider::LOCK_SHARED);
 
+		$event->setArgument('output', $output);
```
create a new event instance, don't  resend the same one

PVince81 commented on apps/files_sharing/tests/ApiTest.php:
```diff
@@ -115,6 +116,10 @@ private function createOCS($request, $userId) {
 			}));
 
+		\OC::$server->getEventDispatcher()->addListener('share.beforeCreate', function (GenericEvent $event) {
+			$event->setArgument('run', true);
```
is this really needed ? "run" is true by default

sharidas replied:
This was added because the tests were failing. It was using false value from Share20OCSTest.php. That's the reason I set `run` value to true here.

PVince81 replied:
Strange. Do you mean one test is influencing another ? This shouldn't happen. We need to find a way to reset the event dispatcher in `tearDown` to avoid test side effects.

PVince81 commented on lib/private/legacy/files.php:
```diff
@@ -139,6 +139,13 @@ public static function get($dir, $files, $params = null) {
+			//Dispatch an event to see if any apps have problem with download
+			$event = new \Symfony\Component\EventDispatcher\GenericEvent(null, ['dir' => $dir, 'files' => $files]);
+			OC::$server->getEventDispatcher()->dispatch('file.beforeCreateZip', $event);
+			if ($event->hasArgument('run') and ($event->getArgument('run') === false) and ($event->hasArgument('message'))) {
```
Make sure the "run" argument is created by you and set to "true" by default like you did for share creation.
Any reason you chose to not simply use the same pattern like for share creation ?

rename "message" to "errorMessage"

PVince81 reviewed (changes requested):
better, still some changes needed

sharidas commented:
Updated the changes:
- made sure that run is set in the files.php
- Created new instance of event for after share event 
- The reason for making run to true is addressed at : https://github.com/owncloud/core/pull/29885#discussion_r158502885

PVince81 commented on lib/private/legacy/files.php:
```diff
@@ -139,6 +139,13 @@ public static function get($dir, $files, $params = null) {
+			//Dispatch an event to see if any apps have problem with download
+			$event = new \Symfony\Component\EventDispatcher\GenericEvent(null, ['dir' => $dir, 'files' => $files, 'run' => true]);
+			OC::$server->getEventDispatcher()->dispatch('file.beforeCreateZip', $event);
+			if (($event->getArgument('run') === false) and ($event->hasArgument('errorMessage'))) {
```
this should be `||`, not `and`. If run is false and no error message was supplied, we should still stop running.

PVince81 commented on lib/private/legacy/files.php:
```diff
@@ -167,6 +174,9 @@ public static function get($dir, $files, $params = null) {
 			$streamer->finalize();
 			set_time_limit($executionTime);
 			self::unlockAllTheFiles($dir, $files, $getType, $view, $filename);
+			$event = new \Symfony\Component\EventDispatcher\GenericEvent(null, ['result' => 'success']);
```
please also pass `dir` and `files` arguments here

sharidas commented:
@PVince81 Let me know if this change looks ok.

sharidas commented:
Created backport to stable10 PR: https://github.com/owncloud/core/pull/30067

PVince81 reviewed (approved):
:+1:
